dev-tools/mage/benchmark.go

```diff
@@ -0,0 +1,170 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package mage
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/elastic/elastic-agent-libs/dev-tools/mage/gotool"
+	"github.com/magefile/mage/mg"
+)
+
+const (
+	goBenchstat = "golang.org/x/perf/cmd/benchstat@v0.0.0-20230227161431-f7320a6d63e8"
+)
+
+var (
+	benchmarkCount = 8
+)
+
+// Benchmark namespace for mage to group all the related targets under this namespace
+type Benchmark mg.Namespace
+
+// Deps installs required plugins for reading benchmarks results
+func (Benchmark) Deps() error {
+	err := gotool.Install(gotool.Install.Package(goBenchstat))
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Run execute the go benchmark tests for this repository, define OUTPUT to write results into a file
+func (Benchmark) Run(ctx context.Context) error {
+	mg.Deps(Benchmark.Deps)
+	log.Println(">> go Test: Benchmark")
+	outputFile := os.Getenv("OUTPUT")
+	benchmarkCountOverride := os.Getenv("BENCH_COUNT")
+	if benchmarkCountOverride != "" {
+		var overrideErr error
+		benchmarkCount, overrideErr = strconv.Atoi(benchmarkCountOverride)
+		if overrideErr != nil {
+			return fmt.Errorf("failed to parse BENCH_COUNT, verify that you set the right value: , %w", overrideErr)
+		}
+	}
+	projectPackages, er := gotool.ListProjectPackages()
+	if er != nil {
+		return fmt.Errorf("failed to list package dependencies: %w", er)
+	}
+	cmdArg := fmt.Sprintf("test -count=%d -bench=Bench -run=Bench", benchmarkCount)
+	cmdArgs := strings.Split(cmdArg, " ")
+	for _, pkg := range projectPackages {
+		cmdArgs = append(cmdArgs, filepath.Join(pkg, "/..."))
+	}
+	_, err := runCommand(ctx, nil, "go", outputFile, cmdArgs...)
+
+	var goTestErr *exec.ExitError
+	switch {
+	case goTestErr == nil:
+		return nil
+	case errors.As(err, &goTestErr):
+		return fmt.Errorf("failed to execute go test -bench command: %w", err)
+	default:
+		return fmt.Errorf("failed to execute go test -bench command %w", err)
+	}
+}
+
+// Diff compare 2 benchmark outputs, Required BASE variable for parsing results, define NEXT to compare base with next and optional OUTPUT to write to file
```

Comment:
The godoc here is a little confusing at first, maybe mention that `BASE` and `NEXT` are environment variables?

Reply:
I tried to write it a bit differently, please let me know if it's better now.